locator/abstract_replication_strategy.cc

```diff
@@ -341,10 +343,195 @@ inet_address_vector_replica_set effective_replication_map::get_natural_endpoints
+}
+
+future<> effective_replication_map::calculate_pending_ranges_for_replacing(
+        const address_ranges& address_ranges,
+        pending_ranges& new_pending_ranges,
+        const std::unordered_map<inet_address, inet_address>& replacing_endpoints) const {
+    if (replacing_endpoints.empty()) {
+        co_return;
+    }
+    for (const auto& node : replacing_endpoints) {
+        auto existing_node = node.first;
+        auto replacing_node = node.second;
+        for (const auto& x : address_ranges) {
+            if (x.first == existing_node) {
+                rslogger.debug("Node {} replaces {} for range {}", replacing_node, existing_node, x.second);
+                new_pending_ranges.emplace(x.second, replacing_node);
+            }
+            co_await coroutine::maybe_yield();
+        }
+    }
+}
+
+future<> effective_replication_map::calculate_pending_ranges_for_leaving(
+        const address_ranges& address_ranges,
+        pending_ranges& new_pending_ranges,
+        const token_metadata& all_left_metadata,
+        const std::unordered_set<inet_address>& leaving_endpoints) const {
+    // get all ranges that will be affected by leaving nodes
+    std::unordered_set<range<token>> affected_ranges;
+    for (auto endpoint : leaving_endpoints) {
+        auto r = address_ranges.equal_range(endpoint);
+        for (auto x = r.first; x != r.second; x++) {
+            affected_ranges.emplace(x->second);
+        }
+        co_await coroutine::maybe_yield();
+    }
+    // for each of those ranges, find what new nodes will be responsible for the range when
+    // all leaving nodes are gone.
+    auto metadata = co_await _tmptr->clone_only_token_map();
+    auto affected_ranges_size = affected_ranges.size();
+    rslogger.debug("In calculate_pending_ranges: affected_ranges.size={} starts", affected_ranges_size);
+    for (const auto& r : affected_ranges) {
+        auto t = r.end() ? r.end()->value() : dht::maximum_token();
+        auto current_endpoints = co_await _rs->calculate_natural_endpoints(t, metadata);
+        auto new_endpoints = co_await _rs->calculate_natural_endpoints(t, all_left_metadata);
+        std::vector<inet_address> diff;
+        std::sort(current_endpoints.begin(), current_endpoints.end());
+        std::sort(new_endpoints.begin(), new_endpoints.end());
+        std::set_difference(new_endpoints.begin(), new_endpoints.end(),
+            current_endpoints.begin(), current_endpoints.end(), std::back_inserter(diff));
+        for (auto& ep : diff) {
+            new_pending_ranges.emplace(r, ep);
+        }
+    }
+    co_await metadata.clear_gently();
+    rslogger.debug("In calculate_pending_ranges: affected_ranges.size={} ends", affected_ranges_size);
+}
+
+future<> effective_replication_map::calculate_pending_ranges_for_bootstrap(
+        pending_ranges& new_pending_ranges,
+        token_metadata& all_left_metadata,
+        const std::unordered_map<token, inet_address>& bootstrap_tokens) const {
```

Comment:
What happens if it fails? Does it leave the inputs in some intermediate state?

Reply:
`bootstrap_tokens` is not changed.
`new_pending_ranges` and `all_left_metadata` may be left in an intermediate state but they are outputs allocated on the caller `effective_replication_map::update_pending_ranges()` stack anyhow.